File: CabFare.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


#importing Libraries
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from ggplot import *
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.tree import export_graphviz


# In[2]:


#set path
os.chdir("/home/zozo/Documents/edwisor/Project/Cab Fare/")


# In[3]:


#check path
os.getcwd()


# In[4]:


#Load Train and Test data
train=pd.read_csv("train_cab.csv")
test=pd.read_csv("test.csv")
train.shape


# In[5]:


test.shape


# In[6]:


train.head()


# In[7]:


#check datatypes of train
train.dtypes


# In[7]:


#convert datatypes
train['fare_amount'].loc[1123]=430
train['fare_amount']=pd.to_numeric(train['fare_amount'])
train['passenger_count']=train['passenger_count'].astype(object)


# ### Missing Value Analysis

# In[8]:


#Create dataframe with missing percentage
missing_val = pd.DataFrame(train.isnull().sum())

#Reset index
missing_val = missing_val.reset_index()

#Rename variable
missing_val = missing_val.rename(columns = {'index': 'Variables', 0: 'Missing_percentage'})

#Calculate percentage
missing_val['Missing_percentage'] = (missing_val['Missing_percentage']/len(train))*100

#descending order
missing_val = missing_val.sort_values('Missing_percentage', ascending = False).reset_index(drop = True)


# In[48]:


missing_val


# In[11]:


#As Passnenger_count is a categorical variable we will use mode to impute it


# In[11]:


train['passenger_count'].value_counts()


# In[9]:


#mode- as 1 has occurred for most no. of times we will replace na with 1 i.e Mode

train['passenger_count'] = train['passenger_count'].fillna(1)


# In[13]:


#Now to impute missing value of fare_amount we will use Mean or Median


# In[10]:


train['fare_amount'].loc[2000]


# In[16]:


#actual value=17.7
#mean=15.0407
#median=8.5
#As Mean is giving better result we will use mean to replace na in fare_amount variable


# In[11]:


#Mean
train['fare_amount'] = train['fare_amount'].fillna(train['fare_amount'].mean())


# In[18]:


#Median
#train['fare_amount'] = train['fare_amount'].fillna(train['fare_amount'].median())


# In[16]:


missing_val


# In[ ]:





# ### Now we will remove outlier location from train data using test dataset

# In[17]:


#Pickup latitude range
print("Range of Pickup Latitude is", (min(train['pickup_latitude']),max(train['pickup_latitude'])))


# In[18]:


#Dropoff latitude range
print("Range of Dropoff Latitude", (min(train['dropoff_latitude']),max(train['dropoff_latitude'])))


# In[19]:


# now lets see what are the boundaries in test dataset 


# In[20]:


#longitude boundary
print("Longitude Boundary in test")
min(test.pickup_longitude.min(), test.dropoff_longitude.min()),max(test.pickup_longitude.max(), test.dropoff_longitude.max())


# In[21]:


#latitude boundary
print("Latitude Boundary in test")
min(test.pickup_latitude.min(), test.pickup_latitude.min()),max(test.pickup_latitude.max(), test.pickup_latitude.max())


# In[12]:


#set boundaries
boundary={'min_longitude':-74.263242,
              'min_latitude':40.573143,
              'max_longitude':-72.986532, 
              'max_latitude':41.709555}


# In[13]:


train.loc[~((train.pickup_longitude >= boundary['min_longitude'] ) & (train.pickup_longitude <= boundary['max_longitude']) &
            (train.pickup_latitude >= boundary['min_latitude']) & (train.pickup_latitude <= boundary['max_latitude']) &
            (train.dropoff_longitude >= boundary['min_longitude']) & (train.dropoff_longitude <= boundary['max_longitude']) &
            (train.dropoff_latitude >=boundary['min_latitude']) & (train.dropoff_latitude <= boundary['max_latitude'])),'outlier']=1
train.loc[((train.pickup_longitude >= boundary['min_longitude'] ) & (train.pickup_longitude <= boundary['max_longitude']) &
            (train.pickup_latitude >= boundary['min_latitude']) & (train.pickup_latitude <= boundary['max_latitude']) &
            (train.dropoff_longitude >= boundary['min_longitude']) & (train.dropoff_longitude <= boundary['max_longitude']) &
            (train.dropoff_latitude >=boundary['min_latitude']) & (train.dropoff_latitude <= boundary['max_latitude'])),'outlier']=0

# Let us drop outlier locations
train=train.loc[train['outlier']==0]
train.drop(['outlier'],axis=1,inplace=True)


# In[14]:


train.shape


# ### Now let us extract important features from pickup_datetime

# In[15]:


#lets create a function to get important features from pickup_datetime variable in train and test datasets
def clean(data):
    data['pickup_datetime']=data.pickup_datetime.str.slice(-23,-3)
    data['pickup_datetime']=pd.to_datetime(data.pickup_datetime)
    data['day'] = data['pickup_datetime'].dt.day
    data['year'] = data['pickup_datetime'].dt.year 
    data['month'] = data['pickup_datetime'].dt.month 
    data['hour'] = data['pickup_datetime'].dt.hour 
    data['weekday'] = data['pickup_datetime'].dt.weekday
    data=data.dropna(subset=['year','month','hour','weekday'])
    return data


# In[16]:


train=clean(train)


# In[17]:


train.shape


# In[18]:


test=clean(test)


# In[19]:


test.shape


# In[ ]:





# ### Now let's calculate trip distance from picup and dropoff latitude and longitude

# In[20]:


def trip_distance(lon1, lat1, lon2, lat2):
    lon1, lat1, lon2, lat2 = map(np.radians, [lon1, lat1, lon2, lat2])
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = np.sin(dlat/2.0)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2.0)**2
    c = 2 * np.arcsin(np.sqrt(a))
    km = 6371 * c  
    return km


# In[21]:


train['trip_distance']=trip_distance(train['pickup_longitude'],train['pickup_latitude'],
                                     train['dropoff_longitude'],train['dropoff_latitude'])


# In[22]:


train.shape


# In[23]:


test['trip_distance']=trip_distance(test['pickup_longitude'],test['pickup_latitude'],
                                     test['dropoff_longitude'],test['dropoff_latitude'])


# In[24]:


test.shape


# In[29]:


train.head()


# In[ ]:





# In[36]:


#Now look at the summary of the data
train.describe()


# #### 1.look at the summary of fare_amount,passenger_count and trip_distance
# #### 2.fare_amount has minimun value as negative value which is not possible so we will drop negative values
# #### 3.passenger_count has minimum value 0 which also does not have any significant value for our model so we will drop it
# #### 4.trip_distance also has minimum value as 0 which is also of no use
# 

# In[40]:


#let's clean fare_amount variable


# In[25]:


#Removing all the fares having value zero
train.drop(train[train['fare_amount'] < 1].index, inplace = True)


# In[26]:


#Removing all the passenger_counts having value zero or more than six
train=train[train['passenger_count']<=6]
train=train[train['passenger_count']>=1]


# In[27]:


#lets check values of passenger_counts
train['passenger_count'].value_counts()


# ### there cannot be 1.3 passenger so lets remove it

# In[28]:


train=train[train['passenger_count']!=1.3]


# In[29]:


##Now remove the trip_distance having value less than 0.2 as most of people will not take a cab for distance below 200 meters
train = train.loc[train['trip_distance'] >=0.2]


# In[68]:


train['trip_distance'].value_counts()


# In[ ]:





# In[43]:


train.describe()


# ###Now if you will carefully look at the summary of data you will find out that max value in fare_amount is 54343
# which is way more than the mean value which is 15.19 that means this is an outlier. Same is the case with trip_distance where mean is 3.4 and max value is 101.09 so lets remove outliers from data

# ###  outliers

# In[48]:


#first look at the relation between trip_distance and fare_amount


# In[36]:


plt.scatter(x=train['trip_distance'],y=train['fare_amount'])
plt.xlabel("Trip Distance")
plt.ylabel("Fare Amount")
plt.ylim(0, 500)
plt.title("Trip Distance vs Fare Amount")


# #### As you can see in scatterplot that fare amount is almost fixed for trip over 80 kM and sometimes fares are very high for very short distance which is basically an outlier.

# In[37]:


# #Plot boxplot to visualize Outliers
get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(train['fare_amount'])


# ###### Now after looking the scatter plot of Fare vs Trip_distance and boxplot of Fare amount we will remove all the fares having value more than 150

# In[30]:


train=train[train['fare_amount']<150]


# ###### Now lets draw boxplot for trip_distance for train and test datasets

# In[47]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(train['trip_distance'])


# In[48]:


get_ipython().run_line_magic('matplotlib', 'inline')
plt.boxplot(test['trip_distance'])


# ##### After visualising boxplot for trip_distance of train and test datasets we know that they contain almost same trip distance so we will leave trip distance as it is.

# 

# In[31]:


train.shape


# In[50]:


train.describe()


# In[56]:


#now you can see after removing ouliers max value and mean of fare_amount has been changed


# In[44]:


# train.to_csv("trained_data.csv", index=False)


# In[45]:


# test.to_csv("processed_test_data.csv",index=False)


# 

# In[51]:


#let's visualize fare_amount
plt.figure(figsize = (14, 4))
n, bins, patches = plt.hist(train.fare_amount, 50, facecolor='green', alpha=0.8)
plt.xlabel('Fare Amount')
plt.title('Histogram of Fare Amount')
plt.xlim(0, 100)
plt.show();


# In[60]:


#as we can see in the histogram most Fare drops between 5 to 15 dollars indicating short trips


# In[52]:


#now lets visualize passenger_count
train['passenger_count'].value_counts().plot.bar(color = 'blue', edgecolor = 'black');
plt.title('Bargraph of passenger counts'); plt.xlabel('Passenger counts'); plt.ylabel('Count');


# In[62]:


#you can see in the above bargraph that most of the time single passengers have booked cab and family booking is least


# In[53]:


#lets visualize the trip_distance\
plt.figure(figsize = (14, 4))
n, bins, patches = plt.hist(train.trip_distance, 200, facecolor='blue', alpha=0.75)
plt.xlabel('trip_distance')
plt.xlim(0, 40)
plt.title('Histogram of trip distance')
plt.show();


# In[64]:


#in the above histogram we can see that maximum no. of times people took short trips between somewhere 0.7 to 3 KM


# In[39]:


train.shape


# In[32]:


def modeling(data,target,drop_cols,split=0.25):
    new_data=data.drop(drop_cols,axis=1)
    X=new_data.drop([target],axis=1)
    y=new_data[target]
    X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=split,random_state=42)
    print("train shape",X_train.shape)
    print("test shape",X_test.shape)
    return X_train, X_test, y_train, y_test
    return new_data


# In[33]:


X_train, X_test, y_train, y_test=modeling(train,'fare_amount',drop_cols=['pickup_datetime'],split=0.2)


# In[57]:


train.dtypes


# In[58]:


train.head()


# In[59]:


test.dtypes


# ## Model Development

# #### Linear Regression

# In[34]:


# Train the model using the training sets
model1 = sm.OLS(y_train, X_train).fit()


# In[35]:


model1.summary()


# In[36]:


y_pred1=model1.predict(X_test)


# #### Lets use RMSE to test accuracy of the Model

# In[37]:


lm_rmse=np.sqrt(mean_squared_error(y_pred1, y_test))
print("RMSE for Linear Regression is ",lm_rmse)


# ##### Now lets try another linear regression by changing split

# In[39]:


X_train1, X_test1, y_train1, y_test1=modeling(train,'fare_amount',drop_cols=['pickup_datetime'],split=0.3)


# In[40]:


# Train the model using the training sets
model2 = sm.OLS(y_train1, X_train1).fit()


# In[41]:


model2.summary()


# In[42]:


y_pred2=model2.predict(X_test1)


# In[43]:


lm_rmse=np.sqrt(mean_squared_error(y_pred2, y_test1))
print("RMSE for Linear Regression is ",lm_rmse)


# In[ ]:





# #### DecisionTree

# In[44]:


#Decision tree for regression
fit_DT = DecisionTreeRegressor(max_depth=6,random_state=42).fit(X_train, y_train)


# In[45]:


predictions_DT = fit_DT.predict(X_test)


# #### Lets use RMSE to test accuracy of the Model

# In[46]:


dt_rmse=np.sqrt(mean_squared_error(predictions_DT,y_test))
print("RMSE = ",dt_rmse)


# In[ ]:





# #### Random Forest

# In[47]:


fit_RF = RandomForestRegressor(n_estimators = 50,random_state=42).fit(X_train,y_train)


# In[48]:


prediction_RF=fit_RF.predict(X_test)


# #### Lets use RMSE to test accuracy of the Model

# In[49]:


rf_rmse=np.sqrt(mean_squared_error(prediction_RF,y_test))
print("RMSE = ",rf_rmse)


# In[ ]:





# ### As we got best Accuracy with RandomForest Model we will use this Model to predict Fare

# In[75]:


test.describe()


# In[76]:


test.head()


# In[77]:


test=test.drop(['pickup_datetime'], axis=1)


# In[78]:


test.shape


# In[79]:


predicted_fare=fit_RF.predict(test)


# In[80]:


test['predicted_fare']=predicted_fare


# In[81]:


test.head(10)


# In[82]:


test.to_csv("test_predicted.csv",index=False)


# In[ ]:

File: cabFare.R
setwd("/home/zozo/Documents/edwisor/Project/Cab Fare")
getwd()

#Load Libraries
x = c("ggplot2", "corrgram", "DMwR", "caret", "randomForest", "unbalanced", "C50", "dummies", "e1071", "Information",
      "MASS", "rpart", "gbm", "ROSE", 'sampling', 'DataCombine', 'inTrees')

#install.packages(x)
lapply(x, require, character.only = TRUE)
rm(x)

## Read the data
train = read.csv("train_cab.csv", header = T)
test=read.csv("test.csv",header = T)

##summary of data
summary(train)
dim(train)
sapply(train, typeof)

### convert datatypes
train[,1] = as.numeric(as.character(train[,1]))
train$passenger_count=as.factor(train$passenger_count)


#####Missing Value Analysis###########
missing_val = data.frame(apply(train,2,function(x){sum(is.na(x))}))
missing_val$Columns = row.names(missing_val)
names(missing_val)[1] =  "Missing_percentage"
missing_val$Missing_percentage = (missing_val$Missing_percentage/nrow(train)) * 100
missing_val = missing_val[order(-missing_val$Missing_percentage),]
row.names(missing_val) = NULL
missing_val = missing_val[,c(2,1)]

ggplot(data = missing_val[1:3,], aes(x=reorder(Columns, -Missing_percentage),y = Missing_percentage))+
     geom_bar(stat = "identity",fill = "red")+xlab("Parameter")+
     ggtitle("Missing data percentage (Train)") + theme_bw()

##As passenger_count is categorical variable we will impute it using mode
###Mode Method
train$passenger_count[is.na(train$passenger_count)] =as.data.frame(mode(train$passenger_count))

#df=train
train=df

#actual value=17.7
#mean=15.014
#median=8.5
train[2001,1]

## we will impute missing value of fare_amount by using mean or median method
####Mean Method
train$fare_amount[is.na(train$fare_amount)] = mean(train$fare_amount, na.rm = T)

####Median Method
# train$fare_amount[is.na(train$fare_amount)] = median(train$fare_amount, na.rm = T)


### Now we will remove outlier location from train data using test dataset

#Pickup latitude range
 min(train$pickup_latitude)
 max(train$pickup_latitude)

 #dropoff latitiude range
 min(train$dropoff_latitude)
 max(train$dropoff_latitude)
 
 #Pickup longitude range
 min(train$pickup_longitude)
 max(train$pickup_longitude)
 
 #dropoff longitude range
 min(train$dropoff_longitude)
 max(train$dropoff_longitude)
 
 # now lets see what are the boundaries in test dataset 
 
 #longitude boundary
 min(test$pickup_longitude, test$dropoff_longitude)
 max(test$pickup_longitude, test$dropoff_longitude)
 # longitude boundary=(-74.26324,-72.98653)
 
 #latitude boundary
 min(test$pickup_latitude, test$dropoff_latitude)
 max(test$pickup_latitude, test$dropoff_latitude)
 #latitude boundary=(40.56897,41.70956)
 
#set boundaries
min_longitude=-74.26324
min_latitude=40.56897
max_longitude=-72.98653
max_latitude=41.70956

train=subset(train, pickup_longitude >= min_longitude)
train=subset(train,pickup_longitude <= max_longitude)
train=subset(train,pickup_latitude >= min_latitude)
train=subset(train,pickup_latitude<=max_latitude)

train=subset(train, dropoff_longitude >= min_longitude)
train=subset(train,dropoff_longitude <= max_longitude)
train=subset(train,dropoff_latitude >= min_latitude)
train=subset(train,dropoff_latitude<=max_latitude)

df=train
df1=test
train=df
#lets create a function to get important features from pickup_datetime variable in train and test datasets
clean=function(data)
{
  data$week_days = weekdays(as.Date(data$pickup_datetime))
  data %>% separate(pickup_datetime, 
                  c("Year","Month", "Day","hour"))
}

train=clean(train)
train=na.omit(train)
train$week_days=sapply(as.character(train$week_days), switch, "Monday" = 1, "Tuesday" = 2, "Wednesday" = 3, "Thursday" = 4,
                       "Friday"=5,"Saturday"=6,"Sunday"=7,USE.NAMES = F)


test=clean(test)
test$week_days=sapply(as.character(test$week_days), switch, "Monday" = 1, "Tuesday" = 2, "Wednesday" = 3, "Thursday" = 4,
                       "Friday"=5,"Saturday"=6,"Sunday"=7,USE.NAMES = F)


### Now let's calculate trip distance from picup and dropoff latitude and longitude
## Haversine
trip_distance = function(lon1, lat1, lon2, lat2){
  # convert decimal degrees to radians
  lon1 = lon1 * pi / 180
  lon2 = lon2 * pi / 180
  lat1 = lat1 * pi / 180
  lat2 = lat2 * pi / 180
  # haversine formula
  dlon = lon2 - lon1
  dlat = lat2 - lat1
  a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
  c = 2 * atan2(sqrt(a), sqrt(1-a))
  km = 6367 * c
  return(km)
}

## Calculating trip_distance for train data
train$trip_distance=trip_distance(train$pickup_longitude,train$pickup_latitude,
                                  train$dropoff_longitude,train$dropoff_latitude)
 
 
## Calculating trip_distance for test data
test$trip_distance=trip_distance(test$pickup_longitude,test$pickup_latitude,
                                  test$dropoff_longitude,test$dropoff_latitude)

## now lets look at the summary of data 
summary(train)
 
#### 1.look at the summary of fare_amount,passenger_count and trip_distance
#### 2.fare_amount has minimun value as negative value which is not possible so we will drop negative values
#### 3.trip_distance also has minimum value as 0 which is also of no use
#### 4.passenger_count has values other than the 1 to 6 so we will remove it

#let's clean fare_amount variable
 #Removing all the fares having value zero
train=subset(train, fare_amount >= 1)

 #Removing all the passenger_counts having value other than one to six
train$passenger_count <- as.numeric(as.character(train$passenger_count))
train=subset(train, passenger_count >= 1 & passenger_count <= 6)
train$passenger_count=as.factor(train$passenger_count)

#lets check values of passenger_counts
table(train$passenger_count)
### there cannot be 1.3 passenger so lets remove it
train=subset(train, passenger_count != 1.3)
length(unique(train$passenger_count))

##Now remove the trip_distance having value less than 0.2 as most of people will not take 
## a cab for distance below 200 meters
train=subset(train, trip_distance>=0.2)


### Now Once again look at the summary of the data
summary(train)

### Now if you will carefully look at the summary of data you will find out that max value in fare_amount is 54343
### which is way more than the mean value which is 15.17 that means this is an outlier. 
### Same is the case with trip_distance where mean is 3.43 and max value is 101.03 so lets remove outliers from data 
 
############# OUTLIERS #####################

#first look at the relation between trip_distance and fare_amount
library("scales") 
ggplot(train, aes_string(x = train$trip_distance, y = train$fare_amount)) + 
  geom_point(colour ="blue") +
  theme_bw()+ ylab("Fare Amount") + xlab("Trip Distance") + ggtitle("Trip Distance vs Fare Amount") + 
  theme(text=element_text(size=10))+ scale_x_continuous(limits = c(0, 100)) + 
  scale_y_continuous(limits = c(0, 500))

#### As you can see in scatterplot that fare amount is almost fixed for trip over 80 kM and sometimes fares are 
### very high for very short distance which is basically an outlier.

# #Plot boxplot to visualize Outliers
#Box plot
qplot(y=train$fare_amount, x= 1, geom = "boxplot",outlier.colour='red')

#### Now after looking the scatter plot of Fare vs Trip_distance and boxplot of Fare amount 
#### we will remove all the fares having value more than 150

train=subset(train, fare_amount < 150)

###### Now lets draw boxplot for trip_distance for train and test datasets

qplot(y=train$trip_distance, x= 1, geom = "boxplot",outlier.colour='red')

qplot(y=test$trip_distance, x= 1, geom = "boxplot",outlier.colour='red')

##### After visualising boxplot for trip_distance of train and test datasets we know that they contain 
##### almost same trip distance so we will leave trip distance as it is. 

# once again have a look at the summary of the data

summary(train)
#now you can see after removing ouliers max value and mean of fare_amount has been changed
 
#let's visualize fare_amount
ggplot(train, aes_string(x = train$fare_amount)) + 
  geom_histogram(fill="green", colour = "black") + geom_density() +
  scale_y_continuous(breaks=pretty_breaks(n=10)) + 
  scale_x_continuous(breaks=pretty_breaks(n=10))+
  theme_bw() + xlab("Fare Amount") + ylab("Frequency") + ggtitle("Histogram of Fare Amount") +
  theme(text=element_text(size=20))+ scale_x_continuous(limits = c(0, 100))
 
# as we can see in the histogram most Fare drops between 5 to 15 dollars indicating short trips
 
### now lets visualize passenger_count
ggplot(train, aes_string(x = train$passenger_count)) +
  geom_bar(stat="count",fill =  "blue") + theme_bw() +
  xlab("Passenger Count") + ylab('Count') + scale_y_continuous(breaks=pretty_breaks(n=10)) +
  ggtitle("Marketing Campaign Analysis") +  theme(text=element_text(size=10))
# you can see in the above bargraph that most of the time single passengers have booked cab and 
# family booking is least

# lets visualize the trip_distance
ggplot(train, aes_string(x = train$trip_distance)) + 
  geom_histogram(fill="green", colour = "black") + geom_density() +
  scale_y_continuous(breaks=pretty_breaks(n=10)) + 
  scale_x_continuous(breaks=pretty_breaks(n=10))+
  theme_bw() + xlab("Fare Amount") + ylab("Frequency") + ggtitle("Histogram of Fare Amount") +
  theme(text=element_text(size=15))
# in this histogram we can see that maximum no. of times people took short trips below somewhere 5KM

############### MODELING #########################
#Divide the data into train and test
set.seed(123)
train_index = sample(1:nrow(train), 0.8 * nrow(train))
train1= train[train_index,]
test1 = train[-train_index,]

### Linear Regression
fit_LM = lm(fare_amount ~ ., data = train1)
#Predict for new test cases
predictions_LM = predict(fit_LM, test1[,-1])

#### Lets use RMSE to test accuracy of the Model
RMSE = function(y, yhat){
  sqrt(mean((y - yhat)^2))
}

RMSE(test1[,1],predictions_LM)
## Accuracy=94.651628
## Error Rate=5.348372


### Decision Tree Regression
# ##rpart for regression
fit_DT = rpart(fare_amount ~ ., data = train1, method = "anova")
predictions_DT = predict(fit_DT, test1[,-1])

#### Lets use RMSE to test accuracy of the Model
RMSE = function(y, yhat){
  sqrt(mean((y - yhat)^2))
}

RMSE(test1[,1],predictions_DT)

## Accuracy=95.560735
## Error Rate=4.439265

### Random Forest Regression

fit_RF <- randomForest(fare_amount ~ .,data    = train1)
predictions_RF=predict(fit_RF,test1[,-1])

#### Lets use RMSE to test accuracy of the Model
RMSE = function(y, yhat){
  sqrt(mean((y - yhat)^2))
}

RMSE(test1[,1],predictions_RF)

### Accuracy=96.333269
### Error Rate=3.666731
 

### As we got best Accuracy with RandomForest Model we will use this Model to predict Fare
summary(test)

### Lets equalize classes of training and test set. Bind the first row of training set 
### to the test set and than delete it
test = rbind(train[1,2:12] , test)
test = test[-1,]
predicted_Fare=predict(fit_RF,test)

test$Predicted_fare=predicted_Fare

write.csv(train, "train_R.csv", row.names = F)
write.csv(test, "predicted_test_R.csv", row.names = F)
